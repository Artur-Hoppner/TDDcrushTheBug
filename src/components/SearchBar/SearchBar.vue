<template>
  <div>
    <input type="text" @keyup="getByThisKeyword($event.target.value)" />
  </div>
</template>
<script>
import { mapActions } from 'vuex';
export default {
  name: 'SearchBar',
  methods: {
    ...mapActions('searchbar', ['getByThisKeyword'])
  },
  mounted() {
    this.getByThisKeyword('');
  }
};
</script>
