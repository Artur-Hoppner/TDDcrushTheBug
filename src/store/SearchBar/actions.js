const actions = {
  getByThisKeyword(context, keyWord) {
    const search = keyWord.toLowerCase();
    if (!search.length) {
      context.rootState.filteredProducts = context.rootState.allProducts;
    } else {
      const foundProducts = context.rootState.allProducts.filter(item =>
        item.match(search)
      );
      context.rootState.filteredProducts = foundProducts;
    }
  },
  displayThisEntireArray(context) {
    context.rootState.filteredProducts = context.rootState.allProducts;
  }
};

export default actions;
