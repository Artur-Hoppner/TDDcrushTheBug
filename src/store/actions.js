const actions = {
  addThisToCart(context, product) {
    context.commit('addToCart', product);
  },
  changeThisProductButtonToggle(context, boolean) {
    context.commit('changeProductButtonToggle', boolean);
  },
  removeThisFromCart(context, id) {
    context.commit('removeFromCart', id);
  },
  increaseThisQuantity(context, product) {
    context.commit('increaseQuantity', product);
  },
  decreaseThisQuantity(context, product) {
    context.commit('decreaseQuantity', product);
  },
  getByThisKeyword(context, keyWord) {
    context.commit('getByKeyword', keyWord);
  },
  displayThisEntireArray(context) {
    context.commit('displayEntireArray');
  }
};

export default actions;
