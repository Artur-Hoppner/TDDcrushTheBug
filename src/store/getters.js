//***********************/
//*** GLOBAL GETTERS ***/
//*********************/

const getters = {
  getProductButtonToggle: state => {
    return state.productButtonToggle;
  },
  getCartProducts: state => {
    return state.cart;
  },
  filteredByKeyword: state => {
    return state.filteredProducts;
  },
  getOrderInfo: state => {
    return state.orderInfo;
  },
  getOrderItems: state => {
    return state.orderInfo.orderList;
  },
  cartIconItems: state => {
    let total = 0;
    state.cart.forEach(cart => {
      total += cart.quantity;
    });
    return total;
  },
  showHamburger: state => {
    return state.isBurgerActive;
  }
};

export default getters;
