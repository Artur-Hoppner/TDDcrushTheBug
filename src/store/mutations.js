//*************************/
//*** GLOBAL MUTATIONS ***/
//***********************/

const mutations = {
  //*************************************/
  //*** CART ARRAY RELATED MUTATIONS ***/
  //***********************************/
  //Add product to cart, takes product object as argument
  addToCart(state, product) {
    //Check if cart already has existing item so no doublettes will be created, only update quantity & price
    let checkExisting = state.cart.filter(item => item.id == product.id);
    if (checkExisting.length >= 1) {
      let index = state.cart.findIndex(item => item.id === product.id);
      state.cart[index].quantity++;
      state.cart[index].price += product.price;
    } else {
      state.cart.push({
        id: product.id,
        tag: product.tag,
        title: product.title,
        desc: product.desc,
        price: product.price,
        quantity: 1
      });
    }
  },
  //Remove Product from cart using index
  removeFromCart(state, id) {
    let findIndex = state.cart.find(item => item.id == id);
    let index = state.cart.indexOf(findIndex);
    state.cart.splice(index, 1);
  },
  //Increase Product Quantity in cart
  increaseQuantity(state, product) {
    //Check for default values of object in original allProducts array
    //Increase Product quantity and price
    let adjustPrice = state.allProducts.find(item => item.id == product.id);
    let index = state.cart.findIndex(item => item.id === product.id);
    state.cart[index].quantity++;
    state.cart[index].price += adjustPrice.price;
  },
  //Decrease Product Quantity in cart
  decreaseQuantity(state, product) {
    //Check for default values of object in original allProducts array
    //Decrease Product quantity and price
    let adjustPrice = state.allProducts.find(item => item.id == product.id);
    let index = state.cart.findIndex(item => item.id === product.id);
    state.cart[index].quantity--;
    state.cart[index].price -= adjustPrice.price;
    if (state.cart[index].quantity == 0) {
      state.cart.splice(index, 1);
    }
  },
  //Create order object based on state.cart
  checkoutCreateOrder(state) {
    if (state.cart != '') {
      //Randomize ordernumber and calculate total cost
      let newOrderNumber = Math.floor(Math.random() * 100000 + 1);
      let totalCost = 0;
      for (let i = 0; i < state.cart.length; i++) {
        totalCost += state.cart[i].price;
      }
      //Create Order object
      state.orderInfo = {
        orderNumber: newOrderNumber,
        orderCost: totalCost,
        deliveryETA: '3',
        orderList: state.cart
      };
    } else {
      return false;
    }
  },
  //*************************************************/
  //*** FILTEREDPRODUCTS ARRAY RELATED MUTATIONS ***/
  //***********************************************/
  //Take input searchword as argument, then use it to filter through allproducts array and match all products including the searchword
  getByKeyword(state, keyWord) {
    const search = keyWord.toLowerCase();
    if (!search.length) {
      state.filteredProducts = state.allProducts;
    } else {
      let foundProducts = [];
      foundProducts = state.allProducts.filter(item => item.tag.match(search));
      state.filteredProducts = foundProducts;
    }
  },
  //Reset the displayArray so it shows all products
  displayEntireArray(state) {
    state.filteredProducts = state.allProducts;
  },
  //*********************************/
  //*** TOGGLE RELATED MUTATIONS ***/
  //*******************************/
  //Change the toggle which decides if ProductItem shows AddToCart button or Increment, Decrement & RemoveFromCart depending on if we are viewing cart or shopping page
  changeProductButtonToggle(state, boolean) {
    state.productButtonToggle = boolean;
  },
  changeCart(state) {
    state.showCart = !state.showCart;
  },
  toggle(state) {
    state.isBurgerActive = !state.isBurgerActive
  }
};

export default mutations;
