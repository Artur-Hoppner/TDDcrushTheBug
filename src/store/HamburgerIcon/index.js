import mutations from './mutations.js';
import getters from './getters.js';
import actions from './actions.js';

const state = {
  namespaced: true,

  state: {
    isBurgerActive: false
  },
  mutations: mutations,
  getters: getters,
  actions: actions
};

export default state;
