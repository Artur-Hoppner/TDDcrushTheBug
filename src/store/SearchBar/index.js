import mutations from './mutations.js';
import getters from './getters.js';
import actions from './actions.js';

const state = {
  namespaced: true,
  //TODO: fix placeholder array with actual products!
  state: {
    allProducts: [
      'cat',
      'dog',
      'parrot',
      'goldfish',
      'horse',
      'elephant',
      'ant',
      'snake',
      'crow',
      'cow',
      'eagle',
      'rhinoceros',
      'chimpanzee'
    ],
    filteredProducts: []
  },
  mutations: mutations,
  getters: getters,
  actions: actions
};

export default state;
