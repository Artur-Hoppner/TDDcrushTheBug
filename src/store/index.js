import Vue from 'vue';
import Vuex from 'vuex';
import searchbar from './SearchBar/index.js';
import actions from './actions.js';
import mutations from './mutations.js';
import getters from './SearchBar/getters.js';
import hamburgerIcon from './HamburgerIcon/index.js'

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    allProducts: [],
    cart: [],
    searchedProducts: []
  },
  actions: actions,
  mutations: mutations,
  getters: getters,
  modules: {
    searchbar: searchbar,
    hamburgerIcon: hamburgerIcon
  }
});
