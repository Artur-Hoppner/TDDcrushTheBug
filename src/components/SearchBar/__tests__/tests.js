import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import store from '../../../store/index.js';
import SearchBar from '@/components/SearchBar/SearchBar.vue';

//create a local instance of our vue
const localVue = createLocalVue();
//make our instance use vuex
localVue.use(Vuex);

describe('User enters a searchword into the searchbar', () => {
  test('Test so that everything renders correctly', () => {
    const wrapper = shallowMount(SearchBar, { store, localVue });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('Test so that an input empty string makes Store.filteredArray full with all products', async () => {
    //Arrange
    const wrapper = shallowMount(SearchBar, { store, localVue });
    const input = wrapper.find('input');
    //Act
    input.setValue('');
    await input.trigger('keyup');
    //Assert
    expect(store.state.filteredProducts.length).toBeGreaterThan(0);
  });

  test('Test so that searchword returns items with matching letters', async () => {
    //Arrange
    const wrapper = shallowMount(SearchBar, { store, localVue });
    const input = wrapper.find('input');
    //Act
    input.setValue('saltgunv3');
    await input.trigger('keyup');
    //Assert
    //Find object containing tag: saltgunv3 in array
    expect(store.state.filteredProducts).toEqual(
      // 1
      expect.arrayContaining([
        // 2
        expect.objectContaining({
          // 3
          tag: 'saltgunv3' // 4
        })
      ])
    );
  });

  test('Test so that searchword with capital letters becomes lowercase', async () => {
    //Arrange
    const wrapper = shallowMount(SearchBar, { store, localVue });
    const input = wrapper.find('input');
    //Act
    input.setValue('SALTGUNV3');
    await input.trigger('keyup');
    //Assert
    //Find object containing tag: saltgunv3 in array
    expect(store.state.filteredProducts).toEqual(
      // 1
      expect.arrayContaining([
        // 2
        expect.objectContaining({
          // 3
          tag: 'saltgunv3' // 4
        })
      ])
    );
  });

  test('Test so that action getByThisKeyword in vuex is called on keyup from component', async () => {
    // Arrange
    const searchbar = {
      namespaced: true,
      state: {},
      mutations: {},
      actions: {
        getByThisKeyword: jest.fn(),
        displayThisEntireArray() {
          return false;
        }
      },
      getters: {}
    };
    const store = new Vuex.Store({
      modules: {
        searchbar
      }
    });
    const wrapper = shallowMount(SearchBar, {
      store,
      localVue
    });

    // Act
    const input = wrapper.find('input');
    await input.trigger('keyup');
    await localVue.nextTick();
    // Assert
    expect(searchbar.actions.getByThisKeyword).toHaveBeenCalled();
  });
});
