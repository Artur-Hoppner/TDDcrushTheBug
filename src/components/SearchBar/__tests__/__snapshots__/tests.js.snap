// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User enters a searchword into the searchbar Test so that everything renders correctly 1`] = `
<div
  id="inputWithIcon"
>
  <input
    id="input"
    placeholder="Search..."
    type="text"
  />
   
  <button>
    Reset
  </button>
</div>
`;
